File: backend/app/schemas/twin.py
"""Twin profile Pydantic schemas."""

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field

from .common import BaseSchema


class TwinGenerateRequest(BaseModel):
    """Request to generate a twin."""

    trigger: Literal["mandatory_modules_complete", "addon_module_complete", "manual"]
    modules_to_include: list[str]


class TwinGenerateResponse(BaseSchema):
    """Response after triggering twin generation."""

    job_id: str
    status: Literal["queued", "processing", "completed", "failed"]
    expected_completion_sec: int
    twin_version: int
    quality_label: Literal["base", "enhanced", "rich", "full"]


class CoverageConfidence(BaseModel):
    """Coverage and confidence for a domain/module."""

    domain: str
    coverage_score: float
    confidence_score: float
    signals_captured: list[str]
    uncertainty_flags: list[str] = Field(default_factory=list)


class OceanEstimate(BaseModel):
    """OCEAN personality trait estimate."""

    score: float
    confidence: float
    evidence: str | None = None


class PersonalityProfile(BaseModel):
    """Personality profile from twin."""

    self_description: str | None = None
    ocean_estimates: dict[str, OceanEstimate] | None = None


class PreferenceDimension(BaseModel):
    """A preference dimension."""

    axis: str
    leaning: str
    strength: float


class BehavioralRule(BaseModel):
    """A behavioral rule extracted from interview."""

    rule: str
    confidence: float
    evidence_turn_ids: list[str] = Field(default_factory=list)


class DecisionProfile(BaseModel):
    """Decision-making profile."""

    speed_vs_deliberation: str | None = None
    gut_vs_data: str | None = None
    risk_appetite: str | None = None
    behavioral_rules: list[BehavioralRule] = Field(default_factory=list)


class StructuredProfile(BaseModel):
    """Full structured profile."""

    demographics: dict[str, Any] = Field(default_factory=dict)
    personality: PersonalityProfile | None = None
    decision_making: DecisionProfile | None = None
    preferences: dict[str, Any] = Field(default_factory=dict)
    communication: dict[str, Any] = Field(default_factory=dict)
    domain_specific: dict[str, Any] = Field(default_factory=dict)
    uncertainty_flags: list[str] = Field(default_factory=list)


class TwinVersionInfo(BaseModel):
    """Information about a twin version."""

    version: int
    modules_included: list[str]
    quality_label: str
    quality_score: float
    created_at: datetime


class TwinProfileResponse(BaseSchema):
    """Full twin profile response."""

    id: UUID
    user_id: UUID
    version: int
    status: Literal["generating", "ready", "failed"]
    modules_included: list[str]
    quality_label: Literal["base", "enhanced", "rich", "full"]
    quality_score: float
    structured_profile: StructuredProfile | None = None
    persona_summary_text: str | None = None
    coverage_confidence: list[CoverageConfidence] = Field(default_factory=list)
    created_at: datetime
    version_history: list[TwinVersionInfo] = Field(default_factory=list)


class TwinListItem(BaseSchema):
    """Twin item for list views."""

    id: UUID
    version: int
    quality_label: str
    quality_score: float
    modules_included: list[str]
    status: str
    created_at: datetime


class AvailableModule(BaseModel):
    """Available add-on module."""

    module_id: str
    module_name: str
    estimated_duration_min: int
    estimated_improvement_percent: str
    description: str

File: frontend/src/types/twin.ts
/**
 * Twin and Chat related TypeScript types.
 */

// Twin Profile types
export interface TwinProfile {
  id: string;
  user_id: string;
  version: number;
  status: 'generating' | 'ready' | 'failed';
  modules_included: string[];
  quality_label: 'base' | 'enhanced' | 'rich' | 'full';
  quality_score: number;
  structured_profile?: StructuredProfile;
  persona_summary_text?: string;
  coverage_confidence: CoverageConfidence[];
  created_at: string;
  version_history: TwinVersionInfo[];
}

export interface StructuredProfile {
  demographics: Record<string, string>;
  personality?: {
    self_description?: string;
    ocean_estimates?: Record<string, { score: number; confidence: number; evidence: string }>;
  };
  decision_making?: {
    speed_vs_deliberation?: string;
    gut_vs_data?: string;
    risk_appetite?: string;
    behavioral_rules?: Array<{ rule: string; confidence: number }>;
  };
  preferences?: Record<string, unknown>;
  communication?: Record<string, string>;
  domain_specific?: Record<string, unknown>;
  uncertainty_flags?: string[];
}

export interface CoverageConfidence {
  domain: string;
  coverage_score: number;
  confidence_score: number;
  signals_captured: string[];
  uncertainty_flags?: string[];
}

export interface TwinVersionInfo {
  version: number;
  modules_included: string[];
  quality_label: string;
  quality_score: number;
  created_at: string;
}

// Twin Generation types
export interface TwinGenerateRequest {
  trigger: 'mandatory_modules_complete' | 'addon_module_complete' | 'manual';
  modules_to_include: string[];
}

// Chat types
export interface TwinChatRequest {
  message: string;
  session_id?: string;
}

export interface EvidenceUsed {
  snippet_id: string;
  why: string;
  snippet_text?: string;
}

export interface TwinChatResponse {
  session_id: string;
  message_id: string;
  response_text: string;
  confidence_score: number;
  confidence_label: 'low' | 'medium' | 'high';
  uncertainty_reason?: string;
  evidence_used: EvidenceUsed[];
  coverage_gaps: string[];
  suggested_module?: string;
}

export interface ChatMessage {
  id: string;
  role: 'user' | 'twin';
  content: string;
  confidence_score?: number;
  confidence_label?: string;
  evidence_used?: EvidenceUsed[];
  coverage_gaps?: string[];
  created_at: string;
}

export interface ChatHistoryResponse {
  session_id: string;
  twin_id: string;
  messages: ChatMessage[];
  created_at: string;
}

export interface ChatSessionItem {
  id: string;
  twin_id: string;
  created_by: string;
  created_at: string;
  message_count: number;
}

export interface BrandChatSessionItem {
  id: string;
  twin_id: string;
  twin_quality_label: string;
  twin_quality_score: number;
  twin_modules: string[];
  created_at: string;
  message_count: number;
}

// Quality label display info
export const QUALITY_LABELS: Record<string, { name: string; color: string; description: string }> = {
  base: {
    name: 'Base',
    color: '#C8FF00',
    description: 'Built from 4 mandatory modules',
  },
  enhanced: {
    name: 'Enhanced',
    color: '#00D4FF',
    description: 'Improved with 1-2 add-on modules',
  },
  rich: {
    name: 'Rich',
    color: '#A855F7',
    description: 'Comprehensive with 3-4 add-on modules',
  },
  full: {
    name: 'Full',
    color: '#F59E0B',
    description: 'Complete profile with all modules',
  },
};

// Confidence display info
export const CONFIDENCE_COLORS: Record<string, { bg: string; text: string; border: string }> = {
  high: { bg: 'bg-green-500/20', text: 'text-green-400', border: 'border-green-500/30' },
  medium: { bg: 'bg-yellow-500/20', text: 'text-yellow-400', border: 'border-yellow-500/30' },
  low: { bg: 'bg-red-500/20', text: 'text-red-400', border: 'border-red-500/30' },
};
